### DISCUSSION TITLE
How can I reference model twice in the different model?

### DISCUSSION THREAD

**Discussion by tbhaxor** [prisma/prisma#23571] [Q&A] [answered]
### Question

So I have a requirement to reference `Account` model twice in the `Project` model

```prisma
model Account {
  id         BigInt       @id @default(autoincrement())
  email      String       @unique @db.VarChar(128)
  name       String       @db.VarChar(256)
  type       AccountType
  createdAt  DateTime     @default(now()) @map("created_at")
  updatedAt  DateTime     @default(now()) @updatedAt @map("updated_at")

  @@map("accounts")
}

model Project {
  id         BigInt        @id @default(autoincrement())
  name       String
  manager    Account       @relation(fields: [managerId], references: [id], onDelete: Cascade)
  managerId  BigInt
  team       Account?      @relation(fields: [teamId], references: [id], onDelete: SetNull)
  teamId     BigInt?
  
  @@map("projects")
}
```

Constraints

- A manager can create multiple projects and may or may not assign it to team.
- A project can have at most 1 team to be assigned at a time.
- Account model should have projects field referencing the projects

I am trying out _connecting table_ approach (idk how you address this). Where I have another model that stores references of all the required model. 

```prisma
model ProjectAccountConnection {
  id        BigInt  @id @default(autoincrement())
  project   Project @relation(fields: [projectId], references: [id], onDelete: Cascade)
  projectId BigInt  @unique
  manager   Account @relation("Manager", fields: [managerId], references: [id], onDelete: Cascade)
  managerId BigInt
  team      Account @relation("Team", fields: [teamId], references: [id], onDelete: Cascade)
  teamId    BigInt
}
```

There is no error in the `Account` model, because of duplicate backrefs.

```prisma
model Account {
  id                       BigInt                     @id @default(autoincrement())
  email                    String                     @unique @db.VarChar(128)
  name                     String                     @db.VarChar(256)
  type                     AccountType
  createdAt                DateTime                   @default(now()) @map("created_at")
  updatedAt                DateTime                   @default(now()) @updatedAt @map("updated_at")
  socialAccounts           SocialAccount[]
+  ProjectAccountConnection ProjectAccountConnection[]
+  ProjectAccountConnection ProjectAccountConnection[]

  @@map("accounts")
}
```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: Linux
- Database: PostgreSQL
- Node.js version: v18.18.2
- Prisma version `prisma -v`: v5.11.0

**Comment by ludralph** [comment, marked as answer]
Hi @tbhaxor 👋 

Thank you for raising this question.

You want to reference the `Account` model twice in the `Project` model. You can do this by defining two separate relations in the `Project` model, one for the `manager` and one for the `team`. Each relation should have a unique name and should reference different fields in the `Project` model. For example:

```prisma
model Account {
  id         BigInt       @id @default(autoincrement())
  email      String       @unique @db.VarChar(128)
  name       String       @db.VarChar(256)
  type       AccountType
  createdAt  DateTime     @default(now()) @map("created_at")
  updatedAt  DateTime     @default(now()) @updatedAt @map("updated_at")
  managedProjects Project[] @relation("Manager")
  teamProjects Project[] @relation("Team")

  @@map("accounts")
}

model Project {
  id         BigInt        @id @default(autoincrement())
  name       String
  manager    Account       @relation("Manager", fields: [managerId], references: [id], onDelete: Cascade)
  managerId  BigInt
  team       Account?      @relation("Team", fields: [teamId], references: [id], onDelete: SetNull)
  teamId     BigInt?
  
  @@map("projects")
}
```
The `Project` model has two relations to the `Account` model: one for the `manager` and one for the `team`. Each relation has a unique name (`Manager` and `Team`) and references a different field in the Project model (`managerId` and `teamId`).
 
On the `Account` side, we have two fields (`managedProjects` and `teamProjects`) that use the @relation attribute to establish a connection to the Project model.

Please note that the `onDelete: Cascade` option means that if an `Account` is deleted, all related `Projects` will also be deleted. The `onDelete: SetNull` option means that if an `Account` is deleted, the `teamId` in the related `Projects` will be set to null.

This approach should meet your constraints: a manager can create multiple projects and may or may not assign it to a team, and a project can have at most one team assigned at a time.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by tbhaxor** [reply]
This is exactly I wanted. However, can we have a conditional relation. Like

```prisma
model Account {
  // Strip other fields...

  type       AccountType
  projects Project[] @relation(type == manager ? "Manager" : "Team")


  @@map("accounts")
}
```

**Reply by ludralph** [reply]
No we can't.
